src/meta/src/barrier/rpc.rs

```diff
@@ -194,16 +204,37 @@ impl ControlStreamManager {
                         .expect("should exist when get collect resp");
                     warn!(node = ?node.worker, err = ?err.as_report(), "get error from response stream");
                     if let Some(command) = node.inflight_barriers.pop_front() {
+                        let errors = self.collect_errors(node.worker.id, err).await;
+                        let err = merge_node_rpc_errors("get error from control stream", errors);
                         self.context.report_collect_failure(&command, &err);
                         break Err(err);
                     } else {
                         // for node with no inflight barrier, simply ignore the error
+                        info!(node = ?node.worker, "no inflight barrier no node. Ignore error");
                         continue;
                     }
                 }
             }
         }
     }
+
+    async fn collect_errors(
+        &mut self,
+        worker_id: WorkerId,
+        first_err: MetaError,
+    ) -> Vec<(WorkerId, MetaError)> {
+        let mut errors = vec![(worker_id, first_err)];
+        let _ = timeout(COLLECT_ERROR_TIMEOUT, async {
+            loop {
+                let (worker_id, result) = self.next_response().await;
+                if let Err(e) = result {
+                    errors.push((worker_id, e));
+                }
```

Comment:
Are we sure that these errors are for the same barrier/epoch?

Reply:
If we only want to resolve the issue, I think it's unnecessary to know the error from the same barrier/epoch, because  before the bidi-stream refactor, on each CN we also didn't distinguish the concrete epoch/barrier where the error comes from. An actor failed at a later epoch will fail the barrier collection of a previous epoch.

If we want to track the detailed epoch that causes the error, we need to maintain the epoch progress per actor, which might require more extra works in some future PRs.

Reply:
Sounds promising. I was just worrying that there could be duplicated error messages from different epochs. However, it seems that we'll fail the entire control stream only once if there's error occurred after the bidi refactor.